Add InvalidateCache to DependencyInjectionNamingSubsystem

generik0 opened the pull request:
Add InvalidateCache to DependencyInjectionNamingSubsystem when services are registering

#552

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/SubSystems/DependencyInjectionNamingSubsystem.cs:
```diff
@@ -123,5 +123,11 @@ public override IHandler GetHandler(Type service)
 			return null;
 		}
 
+		protected override void InvalidateCache()
+		{
+			handlerListsRegistrationOrderByTypeCache.Clear();
+			handlerListsPriorityOrderByTypeCache.Clear();
+			base.InvalidateCache();
+		}
```
Can you add a unit test to show what this is fixing and so we don't get that defect back.

generik0 replied:
Yes, but not until next week.
Good point, Thabks

generik0 commented:
Hi @jonorossi. Tests added. Also cleaned up unused collection. I have changed the handlerListsByTypeCache to be protected, so the overriding GetHandler uses the same collection as the base, And hence the InvalidateCache in the base does not need to be overridden.
Is this as to be expected? @ltines @AntonioDrusin @robertcoltheart @twenzel and thoughts/comments?

twenzel commented on src/Castle.Windsor.Extensions.DependencyInjection.Tests/SubSystems/DependencyInjectionNamingSubsystemTests.cs:
```diff
@@ -0,0 +1,64 @@
+	        Assert.IsType<DecoratedUserService>(actual);
+	        Assert.NotNull(((DecoratedUserService)actual).UserService);
+        }
+
```
Please add following test also:

```csharp
[Fact]
        public void Can_Decorator_Resolve_Decorated_Class_When_Registered_Consecutively()
        {
	        var container = new WindsorContainer();

	        var target = new DependencyInjectionNamingSubsystem();
	        container.Kernel.AddSubSystem(SubSystemConstants.NamingKey, target);

	        container.Register(Component.For<IUserService>().ImplementedBy<DecoratedUserService>() );

 	        container.Register( Component.For<IUserService>().ImplementedBy<UserService>());

	        var actual = container.Resolve<IUserService>();
	        Assert.NotNull(actual);
	        Assert.IsType<DecoratedUserService>(actual);
	        Assert.NotNull(((DecoratedUserService)actual).UserService);
        }
```

generik0 commented:
Thanks @twenzel ypu are welcome to push to my fork if you want.
Otherwise i can fix in a couple of days.

twenzel commented:
@generik0 Sure I can. Can you provide me some permissions.
Let's talk per twitter DM (@twenzel) or Slack (twenzel)

generik0 commented:
Hi @twenzel I don't think the fork is locked is it? Other have been able to push to it. Is you try?
I don't have Twitter, and slack I do not believe you can communicate directly?
You can email and we will sort it all out. rsr@gait.dk

generik0 commented:
@jonorossi if you have time, looks like we are ready to complete :-)
